src/mongodbTypes.ts

```diff
@@ -0,0 +1,185 @@
+      updateOne: Omit<UpdateOneModel<TSchema>, 'filter' | 'update'> & {
+        filter: PaprFilter<TSchema>;
+        update: PaprUpdateFilter<TSchema>;
+      };
+    }
+  | {
+      insertOne: {
+        document: DocumentForInsert<TSchema, TOptions>;
+      };
+    };
+
+export type PaprFilter<TSchema> =
+  | Partial<WithId<TSchema>>
+  | (PaprFilterConditions<WithId<TSchema>> & PaprRootFilterOperators<WithId<TSchema>>);
+
+export type PaprFilterConditions<TSchema> = {
+  [Property in Join<NestedPaths<TSchema, []>, '.'>]?: PaprCondition<
+    PropertyType<TSchema, Property>
+  >;
+};
+
+export interface PaprRootFilterOperators<TSchema> {
+  $and?: PaprFilter<TSchema>[];
+  $nor?: PaprFilter<TSchema>[];
+  $or?: PaprFilter<TSchema>[];
+  $expr?: Record<string, any>;
+  $text?: {
+    $search: string;
+    $language?: string;
+    $caseSensitive?: boolean;
+    $diacriticSensitive?: boolean;
+  };
+  $where?: string | ((this: TSchema) => boolean);
+  $comment?: Document | string;
+}
+
+export type PaprCondition<Type> =
+  | AlternativeType<Type>
+  | PaprFilterOperators<AlternativeType<Type>>;
+
+export interface PaprFilterOperators<TValue> {
+  $eq?: TValue;
+  $gt?: TValue;
+  $gte?: TValue;
+  $in?: readonly TValue[];
+  $lt?: TValue;
+  $lte?: TValue;
+  $ne?: TValue;
+  $nin?: readonly TValue[];
+  $not?: TValue extends string ? PaprFilterOperators<TValue> | RegExp : PaprFilterOperators<TValue>;
+  /**
+   * When `true`, `$exists` matches the documents that contain the field,
+   * including documents where the field value is null.
+   */
+  $exists?: boolean;
+  $type?: BSONType | BSONTypeAlias;
+  $expr?: Record<string, any>;
+  $jsonSchema?: Record<string, any>;
+  $mod?: TValue extends number ? [number, number] : never;
+  $regex?: TValue extends string ? BSONRegExp | RegExp | string : never;
+  $options?: TValue extends string ? string : never;
+  $geoIntersects?: {
+    $geometry: Document;
+  };
+  $geoWithin?: Document;
+  $near?: Document;
+  $nearSphere?: Document;
+  $maxDistance?: number;
+  $all?: TValue extends readonly any[] ? readonly any[] : never;
+  $elemMatch?: TValue extends readonly any[] ? Document : never;
+  $size?: TValue extends readonly any[] ? number : never;
+  $bitsAllClear?: BitwiseFilter;
+  $bitsAllSet?: BitwiseFilter;
+  $bitsAnyClear?: BitwiseFilter;
+  $bitsAnySet?: BitwiseFilter;
+  $rand?: Record<string, never>;
+}
+
+export type PaprMatchKeysAndValues<TSchema> = {
+  [Property in `${NestedPathsOfType<TSchema, any[]>}.$${'' | `[${string}]`}`]?: ArrayElement<
+    PropertyType<TSchema, Property extends `${infer Key}.$${string}` ? Key : never>
+  >;
+} & {
+  [Property in `${NestedPathsOfType<TSchema, Record<string, any>[]>}.$${
+    | ''
+    | `[${string}]`}.${string}`]?: any;
+} & {
+  [Property in Join<NestedPaths<TSchema, []>, '.'>]?: PropertyType<TSchema, Property>;
+};
+
+export interface PaprUpdateFilter<TSchema> {
```

Comment:
I can't help but thinking of extending the mongo type instead of defining a standalone one? What do you think? we can `Omit` or union with whatever we need but I feel, IMO, it would be more descriptive in code of how we diverge from the mongo types...

Reply:
We can't TBH, at least it would be extremely messy, since their usage of `Document` is embedded in a lot of nested types used here.

Reply:
For example, this is the `StrictUpdateFilter` definition in mongodb: https://github.com/mongodb/node-mongodb-native/blob/main/src/mongo_types.ts#L518-L542

We can use it like `Omit<StrictUpdateFilter<TSchema>, Document>`, because `Document` is defined as https://github.com/mongodb/js-bson/blob/6ef100083328b2eeeb3c6a536985f494f919071f/src/bson.ts#L58-L61

Reply:
FWIW, I did that already for `PaprBulkWriteOperation` because it was possible there, since those types weren't merging `Document`.